File: static/js/ChessPieces/Bishop.js
class Bishop extends ChessPiece{
	
	constructor(){
		super();
	}

	movablePlaces(x, y){
		return this.shrinkContinuosArray(this.getBishopMovablePlaces(x, y));
	}

	getBishopMovablePlaces(x, y){
		var matrix = this.live_chessboard_matrix;
		var placeIds = [];
		var leftUp = [];
		var rightUp = [];
		var leftDown = [];
		var rightDown = [];
		if(x>=0&&x<=7&&y>=0&&y<=7){
			for(var i=0; i<=7; i++){
				var b_x = x - i;
				var b_y = y - i;
				if(b_x<=7&&b_x>=0&&b_y<=7&&b_y>=0){
					if(b_x<x&&b_y<y){
						var nextElement = document.getElementById(this.id_gen(b_y, b_x));
						if(nextElement!=null){
							leftUp[i] = nextElement.id;
							if(nextElement.childElementCount!=0){
								break;
							}
						}
					}
				}
			}
			for(var j=0; j<=7; j++){
				var b_x = x + j;
				var b_y = y - j;
				if(b_x<=7&&b_x>=0&&b_y<=7&&b_y>=0){
					if(b_x>x&&b_y<y){
						var nextElement = document.getElementById(this.id_gen(b_y, b_x));
						if(nextElement!=null){
							rightUp[j] = nextElement.id;
							if(nextElement.childElementCount!=0){
								break;
							}
						}
					}
				}
			}
			for(var k=0; k<=7; k++){
				var b_x = x - k;
				var b_y = y + k;
				if(b_x<=7&&b_x>=0&&b_y<=7&&b_y>=0){
					if(b_x<x&&b_y>y){
						var nextElement = document.getElementById(this.id_gen(b_y, b_x));
						if(nextElement!=null){
							leftDown[k] = nextElement.id;
							if(nextElement.childElementCount!=0){
								break;
							}
						}
					}
				}
			}
			for(var n=0; n<=7; n++){
				var b_x = x + n;
				var b_y = y + n;
				if(b_x<=7&&b_x>=0&&b_y<=7&&b_y>=0){
					if(b_x>x&&b_y>y){
						var nextElement = document.getElementById(this.id_gen(b_y, b_x));
						if(nextElement!=null){
							rightDown[n] = nextElement.id;
							if(nextElement.childElementCount!=0){
								break;
							}
						}
					}
				}
			}
		}
		leftUp = this.moveArrayToBack(leftUp);
		rightUp = this.moveArrayToBack(rightUp);
		leftDown = this.moveArrayToBack(leftDown);
		rightDown = this.moveArrayToBack(rightDown);
		var leftToRight = leftDown.concat(rightUp);
		var rightToLeft = rightDown.concat(leftUp);
		placeIds = leftToRight.concat(rightToLeft);
		return placeIds;
	}

	attackingPlaces(isBishop, x, y){
		var attackingBishopPlaces = this.getBishopMovablePlaces(x, y);
		var RightDownAttacking = attackingBishopPlaces[23];
		var LeftUpAttacking = attackingBishopPlaces[31];
		var LeftDownAttacking = attackingBishopPlaces[7];
		var RightUpAttacking = attackingBishopPlaces[15];
		var newAttackingBishopPlaces = [];
		if(isBishop==true){
			this.getAttackingPiecesPlaces(RightDownAttacking, newAttackingBishopPlaces, "comp_bishop", "");
			this.getAttackingPiecesPlaces(LeftUpAttacking, newAttackingBishopPlaces, "comp_bishop", "");
			this.getAttackingPiecesPlaces(LeftDownAttacking, newAttackingBishopPlaces, "comp_bishop", "");
			this.getAttackingPiecesPlaces(RightUpAttacking, newAttackingBishopPlaces, "comp_bishop", "");
		} else {
			this.getAttackingPiecesPlaces(RightDownAttacking, newAttackingBishopPlaces, "", "comp_queen");
			this.getAttackingPiecesPlaces(LeftUpAttacking, newAttackingBishopPlaces, "", "comp_queen");
			this.getAttackingPiecesPlaces(LeftDownAttacking, newAttackingBishopPlaces, "", "comp_queen");
			this.getAttackingPiecesPlaces(RightUpAttacking, newAttackingBishopPlaces, "", "comp_queen");
		}
		return newAttackingBishopPlaces;
	}

	getAttackingPiecesPlaces(placeId, array, type1, type2){
		var type = type1;
		if(type1==""){
			type = type2;
		}
		if(type=="comp_rook"){
		}
		if(placeId!=""){
			var nextPlace = document.getElementById(placeId);
			if(nextPlace.childElementCount!=0){
				var nextPiece = nextPlace.firstElementChild.id;
				if(this.isType(nextPiece, type)){
					array.push(placeId);
				}
			}
		}
	}
}

File: chess_app/ChessEngine/ChessPieces/Bishop.py
from ChessPiece import ChessPiece
from chess_app.ChessEngine.Selecter import Select 

class Bishop(ChessPiece):
	
	def __init__(self):
		super(Bishop, self).__init__()

	def movablePlaces(self, x, y):
		return self.shrinkContinuosArray(self.getBishopMovablePlaces(x, y))

	# gets all the positions a bishop can move gien on coordinates the board 
	def getBishopMovablePlaces(self, x, y):
		matrix = self.live_chessboard_matrix
		placeIds = []
		leftUp = []
		rightUp = []
		leftDown = []
		rightDown = []
		
		if x >= 0 and x <= 7 and y >= 0 and y <= 7:
			
			for i in range(8):
				b_x = x - i
				b_y = y - i
				
				if b_x <= 7 and b_x >= 0 and b_y <= 7 and b_y >= 0:
					if b_x < x and b_y < y:
						rightDownSelect = Select()
						nextElement = rightDownSelect.selectFromParentId(matrix, self.id_gen(b_y, b_x))
						
						if nextElement.parent_id != None and not self.isType(nextElement.piece_id, "comp_"):
							rightDown.append(nextElement.parent_id)
							if nextElement.piece_id != "" and nextElement.piece_id != None:
								break
						if self.isType(nextElement.piece_id, "comp_"):
							break
									
					
			for j in range(8):
				b_x = x + j
				b_y = y - j
				
				if b_x <= 7 and b_x >= 0 and b_y <= 7 and b_y >= 0:
					if b_x > x and b_y < y:
						leftDownSelect = Select()
						nextElement = leftDownSelect.selectFromParentId(matrix, self.id_gen(b_y, b_x))
						
						if nextElement.parent_id != None and not self.isType(nextElement.piece_id, "comp_"):
							leftDown.append(nextElement.parent_id)
							if nextElement.piece_id != "" and nextElement.piece_id != None:
								break
						if self.isType(nextElement.piece_id, "comp_"):
							break
							
			for k in range(8):
				b_x = x - k
				b_y = y + k
				
				if b_x <= 7 and b_x >= 0 and b_y <= 7 and b_y >= 0:
					if b_x < x and b_y > y:
						rightUpSelect = Select()
						nextElement = rightUpSelect.selectFromParentId(matrix, self.id_gen(b_y, b_x))
						
						if nextElement.parent_id!= None and not self.isType(nextElement.piece_id, "comp_"):
							rightUp.append(nextElement.parent_id)
							if nextElement.piece_id != "" and nextElement.piece_id != None:
								break
						if self.isType(nextElement.piece_id, "comp_"):
							break						
					
			for n in range(8):
				b_x = x + n
				b_y = y + n
				
				if b_x <= 7 and b_x >= 0 and b_y <= 7 and b_y >= 0:
					if b_x > x and b_y > y:
						leftUpSelect = Select()
						nextElement = leftUpSelect.selectFromParentId(matrix, self.id_gen(b_y, b_x))
						
						if nextElement.parent_id!= None and not self.isType(nextElement.piece_id, "comp_"):
							leftUp.append(nextElement.parent_id)
							if nextElement.piece_id != "" and nextElement.piece_id != None:
								break
						if self.isType(nextElement.piece_id, "comp_"):
							break
					
		leftUp = self.moveArrayToBack(leftUp)
		rightUp = self.moveArrayToBack(rightUp)
		leftDown = self.moveArrayToBack(leftDown)
		rightDown = self.moveArrayToBack(rightDown)
		# print("leftUp %s rightUp %s leftDown %s rightDown %s"%(leftUp, rightUp, leftDown, rightDown)) 
		leftToRight = leftDown + rightUp
		rightToLeft = rightDown + leftUp
		placeIds = leftToRight + rightToLeft
		
		return placeIds
	

	# sees if a bishop or queen is attacking the position of the coordinates given	
	def attackingPlaces(self, isBishop, x, y):
		attackingBishopPlaces = self.getBishopMovablePlaces(x, y)
		# print("bishop places: %s"%attackingBishopPlaces)
		RightDownAttacking = attackingBishopPlaces[23]
		LeftUpAttacking = attackingBishopPlaces[31]
		LeftDownAttacking = attackingBishopPlaces[7]
		RightUpAttacking = attackingBishopPlaces[15]
		newAttackingBishopPlaces = []
		if isBishop == True:
			self.getAttackingPiecesPlaces(RightDownAttacking, newAttackingBishopPlaces, "player_bishop", "")
			self.getAttackingPiecesPlaces(LeftUpAttacking, newAttackingBishopPlaces, "player_bishop", "")
			self.getAttackingPiecesPlaces(LeftDownAttacking, newAttackingBishopPlaces, "player_bishop", "")
			self.getAttackingPiecesPlaces(RightUpAttacking, newAttackingBishopPlaces, "player_bishop", "")
		else: 
			self.getAttackingPiecesPlaces(RightDownAttacking, newAttackingBishopPlaces, "", "player_queen")
			self.getAttackingPiecesPlaces(LeftUpAttacking, newAttackingBishopPlaces, "", "player_queen")
			self.getAttackingPiecesPlaces(LeftDownAttacking, newAttackingBishopPlaces, "", "player_queen")
			self.getAttackingPiecesPlaces(RightUpAttacking, newAttackingBishopPlaces, "", "player_queen")
		
		return newAttackingBishopPlaces
	

	# attcking pieces helper
	def getAttackingPiecesPlaces(self, placeId, array, type1, type2):
		matrix = self.live_chessboard_matrix
		piece_type = type1
		if type1 == "":
			piece_type = type2
		
		if placeId != "":
			typeSelect = Select()
			nextPlace = typeSelect.selectFromParentId(matrix, placeId)
			
			if nextPlace.piece_id != None and nextPlace.piece_id != "":
				nextPiece = nextPlace.piece_id
				if self.isType(nextPiece, piece_type):
					array.append(placeId)
				
